tools/lsp/preview.rs

```diff
@@ -219,17 +222,85 @@ struct PreviewState {
 }
 thread_local! {static PREVIEW_STATE: std::cell::RefCell<PreviewState> = Default::default();}
 
+pub fn set_design_mode(sender: crate::ServerNotifier, enable: bool) {
+    let mut cache = CONTENT_CACHE.get_or_init(Default::default).lock().unwrap();
+    cache.design_mode = enable;
+
+    configure_design_mode(enable, &sender);
+    send_notification(
+        &sender,
+        if enable { "Design mode enabled." } else { "Design mode disabled." },
+        Health::Ok,
+    );
+}
+
+fn show_document_request_from_element_callback(
+    url: &str,
+    start_line: u32,
+    start_column: u32,
+    end_line: u32,
+    end_column: u32,
+) -> lsp_types::ShowDocumentParams {
+    use lsp_types::{Position, Range, ShowDocumentParams, Url};
+
+    let start_pos = Position::new(start_line - 1, start_column);
+    let end_pos = Position::new(end_line - 1, end_column);
+    let selection = Some(Range::new(start_pos, end_pos));
+
+    let uri = Url::parse(url).unwrap_or_else(|_| Url::parse("file:///invalid").unwrap());
```

Comment:
If we can't parse the URL, we probably should not send the show document notification